common/src/main/java/io/pravega/common/concurrent/Futures.java

```diff
@@ -228,6 +229,40 @@
         return result;
     }
 
+    /**
+     * Returns a CompletableFuture that will complete with the same outcome or result as the given source, but when
+     * cancelled, will apply a consumer to the eventual result of the original future.
+     * <p>
+     * If the returned CompletableFuture is NOT cancelled ({@link CompletableFuture#cancel}):
+     * - If source completes normally, the result CompletableFuture will complete with the same result.
+     * - If source completes exceptionally, the result CompletableFuture will complete with the same result.
+     * <p>
+     * If the returned CompletableFuture is cancelled ({@link CompletableFuture#cancel}):
+     * - If the source has already completed, the result CompletableFuture will also be completed with the same outcome.
+     * - If the source has not already been completed, if it completes normally, then `onCancel` will be applied to
+     * the result when it eventually completes. The source completes exceptionally, nothing will happen.
+     *
+     * @param source   The CompletableFuture to wrap.
+     * @param onCancel A Consumer to invoke on source's eventual completion result if the result of this method is cancelled.
+     * @param <T>      Result type.
+     * @return A CompletableFuture that will complete with the same outcome or result as the given source.
+     */
+    public static <T> CompletableFuture<T> cancellableFuture(CompletableFuture<T> source, Consumer<T> onCancel) {
+        if (source == null) {
+            return null;
+        }
+
+        val result = new CompletableFuture<T>();
+        source.thenAccept(result::complete);
+        Futures.exceptionListener(source, result::completeExceptionally);
```

Comment:
These listeners are on different futures. The first one is on the source (which we're passed in) and propagates the failure to the result. The second one is on the result and invokes the `onCancel` callback if someone cancelled the result. I cannot combine them into one.

Also, these are not affecting the future chain. We are returning `result` directly, yet we are attaching these listeners to the result but not chain them anywhere. They will only be executed in case of some exception; they will not affect the normal path.

Reply:
I see my phrasing was confusing. I was refering to this and the above line.

Reply:
Got it now. Fixed it.